fabric: Introduce peer provider CQ and SRX

shefty opened the pull request:
Define mechanism for sharing a CQ between providers.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

wzamazon commented on include/rdma/fi_ext.h:
```diff
@@ -124,6 +125,32 @@ struct fid_mem_monitor {
+
+struct fi_ops_cq_progress {
+	size_t	size;
+	void	(*progress)(struct fid_peer_cq *cq);
```
Maybe it is better for progress to return number of CQ entry it has written? Sometime we want to keep the progress engine running until something has been completed.

wzamazon replied:
I see. Yes, it make sense. The CQ exporter can keep a counter in its CQ.

wzamazon commented:
Thanks! The CQ structure looks good to me.

As previously discussed, I think we need a shared receive queue to accompany with it. For the CQ importer to be able to claim RX buffer and write RX completion.

shefty commented:
The shared receive context is my next patch.

shefty commented on man/fi_peer.3.md:
```diff
@@ -0,0 +1,162 @@
+etc.) should be set to return -FI_ENOSYS when called.
+
+The importer is responsible for setting up the callbacks with the exporter.
+It does this by calling fi_export_fid(), passing in the input cq passed to
```
The importer is exporting the fid from the peer.  The fi_import_fid / fi_export_fid APIs are already defined.  We need to use fi_export_fid() as that allows returning the new object.
```
fi_export_fid(struct fid *fid, uint64_t flags,
	      struct fid **expfid, void *context);
fi_import_fid(struct fid *fid, struct fid *expfid, uint64_t flags)
```
We could create a new wrapper, say, fi_get_peer_fid()

wzamazon replied:
Yes. I also found it a little confusing

But `fi_import_fid` is being used by open-mpi:

https://github.com/open-mpi/ompi/blob/dfb9ca0bb5d253366512a5fabc11efae9c58f57c/opal/mca/common/ofi/common_ofi.c#L172

j-xiong replied:
A full sequence could explain it better. Here is my understanding:

exporter:
(1) call fi_cq_open() with FI_PEER_EXPORT flag, output cq_fid;
(2) call fi_export_fid() with cq_fid as input, output cq_exfid with export_ops set, also associate cq_fid with cq_exfid.

importer:
(1) call fi_cq_open() with FI_PEER_IMPORT flag, input cq_exfid, output cq_fid2. associate cq_fid2 with cq_exfid, set import_ops.

The importer needs not to know cq_fid from exporter. It should only deal with the exported fid.

j-xiong replied:
Alternatively, the importer should be able to call fi_import_fid() with cq_exfid as the input and local cq_fid as the output instead of calling fi_cq_open(). This match how fi_import_fid() is used for the memory monitor case.

j-xiong replied:
The latter lacks the ability to pass cq_attrs so is only applicable is all default setting are used.

wzamazon replied:
I think there need to be an `getattr` method, so importer can get the attribute of the CQ.

shefty replied:
Updated man page to reword things in places and add a call sequence.  I may update it again to define 3 roles - the exporter, the importer, and the linking entity.  But will wait for other comments.

shefty replied:
The importer is passed attributes through the fi_cq_open() call.  This way, the attributes of the importer can be independent from the attributes of the exporter.  It's hard to imaging how most of the attributes would be used by the importer, but I wanted them there.  At least the cq_attr::format is useful.  Knowing if the exporter is supporting blocking calls might be.

Something not captured by the changes, but that I considered, is that the importer may still need to handle the fi_trywait() call.  But that also gets into sharing fd's (or some other wait object) between the providers.  And since we're really only targeting the shm provider as a peer for now, I decided to ignore this problem.

j-xiong commented on man/fi_peer.3.md:
```diff
@@ -0,0 +1,194 @@
+5. shm allocates a cq, but configures it such that all completions
+   are written to a peer cq.  The cq ops to read from the cq are
+   set to enosys calls.
+6. shm calls fi_export_fid(rxm_cq, &peer_cq, ...) to export the peer
+   cq from the rxm provider.
+7. rxm allocates, initializes, and returns the peer_cq.  rxm creates
+   a reference between the peer_cq object and its rxm_cq object.
```
I still think it is more intuitive to make these two changes:
(1) move 6, 7 before 4 and let rxm call fi_export_fd; 
(2) change 4 to pass peer_cq instead of rxm_cq.

Nothing else needs to change.

shefty replied:
I'm not following how to move 6,7 before 4.  Something must call into shm before it calls fi_export_fid.

We also need to define how the object lifetimes work.  In the model above the shm cq takes a reference on the rxm cq (done via fi_export_fid).  That reference is held until the the peer_cq is closed.  The peer_cq would be closed when closing the shm cq.

We need: Determine which provider owns the allocation of peer_cq.  Which way the object references go.  When the references are taken/released.

j-xiong replied:
Why fi_export_fid needs to be called by shm? rxm can (and should) do it.  With this change step 6 and 7 has nothing to do with shm  and can be moved before 4. 

Calling fi_export_fid from rxm also takes reference of rxm_cq. That doesn't change. The reference is held by the peer_cq. The close method of peer_cq should release the reference.

shefty replied:
Moving 7 before 4, pass peer_cq at 5, then remove 6 should be the same overall flow.  There would be no calls to export/import, as that is assumed by the IMPORT flag.

j-xiong replied:
Yes, but I though 7 was just explaining the internal activities of 6.

shefty replied:
Ah, I rememeber.  The fi_cq_open() call takes a fid_cq ** as the input.  The fid_peer_cq is a different structure.  To make the simplified flow work, the fi_cq_attr would need to be extended (most reasonable path) or use the context field.

shefty replied:
Whatever we come up with here should also be usable with shared rx, meaning we have this:
```
fi_srx_context(struct fid_domain *domain, struct fi_rx_attr *attr,
	       struct fid_ep **rx_ep, void *context)
```
function to work with.  We could do something like:
```
struct fi_peer_cq_attr {
    struct fi_cq_attr attr;
    struct fid_peer_cq *peer_cq;
};
```
(same with fi_rx_attr) and use the flag to know that the structure can be cast.  It would need to be more complex than this to handle forward/backward compatibility changes to the base attr structure.

wzamazon replied:
Do we need to specify `ep` to created shared rx context?

I thought rx context will be shared by multiple end points.

j-xiong replied:
@shefty Yes I like the idea of extending the attrs structures. 
@wzamazon I don't  think ep as an input is needed for creating shared rx context. shared context is associated with endpoint via fi_ep_bind().

shefty commented:
Reworked and simplified the definition.  We only need to define how to import objects.  I defined a new context structure for this and pass in the peer_cq through the ib_cq_open() context parameter, rather than the attributes.  (The attributes are more difficult to deal with, since those can change separately from the import model.)

j-xiong commented on man/fi_peer.3.md:
```diff
@@ -0,0 +1,184 @@
+#include <rdma/fi_ext.h>
+
+int fi_export_fid(struct fid *fid, uint64_t flags,
+	struct fid **expfid, void *context);
```
alignment seems to be off here.

j-xiong commented:
I think the peer cq work flow in this revision is much clearer. 

Now it seems that fi_export_fid/fi_import_fid are not applicable to shared CQ (and maybe share rx context too). Is it possible to define some rules to distinguish these two models (fi_export/import_fid vs. direct use of fi_peer_xxx objects)?

shefty commented:
We have 0 uses of fi_export_fid, 1 for fi_import_fid, and 1 (soon to be 2) for fi_peer_xxx objects.  That's not really enough samples to define rules.  Especially since #6525 may be introducing yet another option.  :)

I can remove the documentation for fi_export_fid, since there isn't any support for it, and update the fi_import_fid documentation to describe the one use case that it does have.

j-xiong commented:
Leaving fi_export_fid() there is fine. We may find its usage in the future.

The layout of the current man page gives the impression that fi_exportd_fid/fi_import_id is the main interface, and the shared CQ is just a special case. Should the fi_peer_xxx objects be put under the `NAME` and `SYNOPSIS` section to clearly show that they are parallel concepts?

shefty commented:
I moved the documentation for export/import fid to the bottom.  I left export, but listed it as reserved for future use.  The peer CQ section is at same level as name and description. I also just added the peer_cq structure in the name/synopsis sections.

shefty commented:
Updated to change wording that I hope makes things clearer.  Replaced exporter/importer syntax with owner/peer.  The owner is the provider that allocates the CQ.  The peer is the user of that CQ.

Added documentation directly related to the callbacks.

tschuett commented:
> Updated to change wording that I hope makes things clearer. Replaced exporter/importer syntax with owner/peer. The owner is the provider that allocates the CQ. The peer is the user of that CQ.

In my area, they are called producer and consumer.

shefty commented:
Added a patch that introduces the peer shared receive context.  That's a lot of qualifiers for the word context, but here we are.  :)  I'm sure I'll tweak the wording in the man page more.

Producer/consumer has a similar problem that exporter/importer has.  Both providers produce/export something and consume/import something else.  I went with owner because only one provider allocates and manages the object's lifetime.  We can still work on the wording here, as long as the proposed functions and data structures work.

j-xiong commented on man/fi_peer.3.md:
```diff
@@ -192,6 +192,120 @@ on the local node.
+
+/* Castable to dlist_entry */
+struct fi_peer_rx_entry {
+    struct fi_peer_rx_entry *next;
+    struct fi_peer_rx_entry *prev;
```
why not use dlist_entry here directly?

shefty replied:
This is exposed with the public API.  dlist_entry is internal, and I don't think we want to define and expose a linked-list API from libfabric.

j-xiong replied:
That makes sense. Thanks.

shefty replied:
I debated whether to include this, since with the tcp provider it actually uses a single-linked list.  shm and rxm use a double-linked list.  But without something here, the owner would need to allocate a structure to track unexpected messages, but the peer already has this.  I went with this as the compromise.

wzamazon commented on man/fi_peer.3.md:
```diff
@@ -192,6 +192,120 @@ on the local node.
+be discarded.
+
+## fi_ops_srx_peer::start_msg() / start_tag()
+
```
If I understand correctly, these two functions are implemented by peer and are called by owner. Right?

shefty replied:
Correct.  The owner would call these in response to the application posting a matching receive.

wzamazon replied:
So the workflow is:

1. application call `fi_recv/fi_trecv` on owner.
2. owner save the receive buffer in a queue
3. Meanwhile, peer receive a message, and need to find a receive buffer for it.
4. Peer call `owner->get_msg_entry/get_tag_entry`.
5. If there are receive in queue, `get_msg_entry` will return the buffer, and return 0. Otherwise, `get_msg/tag_entry` will put the info in an unexpected msg list, and return -FI_EINPROGRESS.
6. when application put in new receive, it first get matched with unexpected msg list, and owner call `peer->start_msg`

I guess a graph would be helpful.

wzamazon replied:
`start_msg` is not easy to understand. May be `start_recv_msg`?

Also, can we put the `peer_rx_entry` as argument? since owner already has that information.

shefty replied:
The flow is correct.  This is an operation of the shared receive context -- 'recv' is a given.  The peer_rx_entry is an argument.  The arguments are just not listed as part of the heading in the man page.

wzamazon commented on man/fi_peer.3.md:
```diff
@@ -192,6 +192,120 @@ on the local node.
+The ownership of structure field values and callback functions is similar
+to those defined for peer CQs, relative to owner versus peer ops.
+
+## fi_ops_srx_owner::get_msg_entry() / get_tag_entry()
```
I think `get` is too general. Maybe a different word like `claim`?

shefty replied:
Claim has a specific meaning relative to the main API.  We're either getting a receive buffer (expected case) or queuing a request for a receive buffer (unexpected case).

wzamazon commented on man/fi_peer.3.md:
```diff
@@ -192,6 +192,120 @@ on the local node.
+of the fi_peer_rx_entry is returned to the peer provider and has been
+initialized for receiving the incoming message.
+
+## fi_ops_srx_peer::discard_msg() / discard_tag()
```
Do we need `peek`?

shefty replied:
No, that should not be needed.  The owner has all the information it needs to handle a peek operation from the app.

wzamazon reviewed (commented):
Can you update the title of the PR to include `SRX`?

wzamazon commented:
How does the peer get the SRX? through `fi_ep_bind`?

If a peer is bind to a SRX, its own recv will not work. right?

shefty commented:
For both the CQ and SRX cases, the peer is provided the owner's object through the fi_cq_open() or fi_srx_context() call, respectively.  The owner object is passed as the context parameter.  It is correct that the peer's srx should fail any call that attempts to post a receive buffer to it.

wzamazon commented:
> fi_srx_context() call, respectively. The owner object is passed as the context parameter. It is correct that the peer's srx should fail any call that attempts to post a receive buffer to it.

I did not see the definition of `fi_srx_context()`?

j-xiong commented:
`fi_srx_context` is an existing API call. See `man fi_endpoint`.

shefty commented:
Added example flows for creating the srx and handling expected and unexpected messages on the srx.

shefty commented:
Added pointer to desc array in rx_entry.  The iov is left as an actual array.  The desc is a pointer and must be set to the array, or NULL in the case of shm.  Renamed iov_cnt to count, since it applies to both arrays

wzamazon commented:
> The desc is a pointer and must be set to the array, or NULL in the case of shm. 

shm will need `desc` if iov is on device.

wzamazon commented on include/rdma/fi_ext.h:
```diff
@@ -124,6 +125,36 @@ struct fid_mem_monitor {
+	size_t	size;
+	ssize_t (*write)(struct fid_peer_cq *cq, void *context, uint64_t flags,
+			size_t len, void *buf, uint64_t data, uint64_t tag,
+			fi_addr_t src);
```
Which address will `src` be?

Note that there are two AVs (one is owner's, the other is peer's), so the same remote endpoint will have two addresses, one for each AV.

This function is called by peer, which write directly to the application facing CQ. so the address `src` should be the address in owner's AV. On the other hand, peer does not know that address. it only knows the address in peer AV.

shefty replied:
As documented, this is the address relative to the peer.  The owner is responsible for translating it if that is needed.  The details are implementation specific.

But, conceptually, there will be 3 "AVs" in use here.  The top level 'AV' translates the user's fi_addr_t value to the local/remote provider and corresponding fi_addr_t' value.  The local/remote providers each have their own AV, which may store different address formats.  The top level 'AV' will need a reverse lookup mechanism.

bwbarrett replied:
My understanding from talking to Wei is that this is largely a performance improvement plan.  Is 3 AVs really an acceptable solution?  If we can get it to single pointer chase, maybe that's not horrible (although it is another cache miss in the average case).  But if there's a search or even a table indirection, we're really starting to increase latency for software abstractions.

shefty replied:
This is targeted as an improvement over the application accessing the 2 (or more) providers independently.  The 3 AV's mentioned above is conceptual, the implementation doesn't need to match that.  But my guess is that this is how the link provider is being implemented.  If an app uses 2 providers independently, then there's conceptually a 3rd 'AV' at the app layer.

One issue is the address formats aren't guaranteed to be the same, and one of the AVs could be implemented in HW.  With rxm, and I think with shm, there's no lookup.  The src_addr is carried in the message or retrieved directly from the connection the message arrived on.  Pairing rxm with shm would require rxm to have some reverse lookup table to convert the shm fi_addr value to some index in the rxm AV.  That table only needs to be sized relative to the number of local peers, so maybe 8-96 entries today.

To take this further, we would introduce some sort of user context that could be stored with each address. 
 It would probably work okay for SW based AV's, but HW AV's might struggle.  Such a feature would also impact shared AVs and AV_MAP options.

Maybe a peer AV can be defined with this feature, which would essentially move the reverse lookup into the shm AV.

wzamazon replied:
> Maybe a peer AV can be defined with this feature, which would essentially move the reverse lookup into the shm AV.

It would be great if one can specify the target fi_address when calling `fi_av_insert` on shm AV. In that case, the owner would generate the fi_address for an remote endpoint, then call `fi_av_insert` on shm av with that address. Then everyone will be using same address for same remote endpoint.

shefty replied:
That potentially forces shm (or whatever provider) to create a huge AV that's almost entirely empty.  It can also only work for if all providers are using AV_TABLE.  What's desired here is the ability to store a context with each address and return that instead.  That context could be another address or pointer.

I think there's an issue in that in a general sense, there's not necessarily a single AV associated with a CQ.  Apps are expected to figure out what AV the src_addr might be referring to based on the completion context.  But the owner doesn't have that data here.

wzamazon commented on include/rdma/fi_ext.h:
```diff
@@ -155,6 +155,54 @@ struct fi_peer_cq_context {
+struct fi_peer_rx_entry {
+	struct fi_peer_rx_entry *next;
+	struct fi_peer_rx_entry *prev;
+	fi_addr_t addr;
```
Which address should it be?
The address in peer's AV or the address in owner's AV.

shefty replied:
Address is relative to the caller (peer provider).  That's all the information that it has.

shefty commented:
These changes are not complete, but there are developers needing the data structures in this patch.  I'm aware of at least one change needed to the documentation, and we will need to resolve the addressing.  Unless there are serious objections, I would like to merge these changes in order to move forward.  I will submit new patches to address issues that have been raised, as well any other problems that are identified once the implementation moves forward.

wzamazon commented:
Sure. Can we have an issue to discuss address?

shefty commented:
Opened #7658 to track the fi_addr_t issue.  The other problem I mentioned above regarding an update to the man pages isn't actually a problem.  The documentation accidently defines the desired behavior.  I mean it intentionally defines the desired behavior, because it was totally planned.  Yeah.  (It was about which provider initialized the fid for the shared object.)
